// @microsoft/teamsfx-core/build/plugins/resource/spfx/utils/questions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.webpartDescriptionQuestion = exports.webpartNameQuestion = exports.frameworkQuestion = exports.SPFXQuestionNames = void 0;
var SPFXQuestionNames;
(function (SPFXQuestionNames) {
    SPFXQuestionNames["framework_type"] = "spfx-framework-type";
    SPFXQuestionNames["webpart_name"] = "spfx-webpart-name";
    SPFXQuestionNames["webpart_desp"] = "spfx-webpart-desp";
})(SPFXQuestionNames = exports.SPFXQuestionNames || (exports.SPFXQuestionNames = {}));
exports.frameworkQuestion = {
    type: "singleSelect",
    name: SPFXQuestionNames.framework_type,
    title: "Framework",
    staticOptions: [
        { id: "react", label: "React" },
        { id: "minimal", label: "Minimal" },
        { id: "none", label: "None" },
    ],
    placeholder: "Select an option",
    default: "react",
};
exports.webpartNameQuestion = {
    type: "text",
    name: SPFXQuestionNames.webpart_name,
    title: "Web Part Name",
    default: "helloworld",
    validation: {
        pattern: "^[a-zA-Z_][a-zA-Z0-9_]*$",
    },
};
exports.webpartDescriptionQuestion = {
    type: "text",
    name: SPFXQuestionNames.webpart_desp,
    title: "Web Part Description",
    default: "helloworld description",
    validation: {
        required: true,
    },
};
