// @microsoft/teamsfx-core/build/plugins/resource/bot/question.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.showNotificationTriggerCondition = exports.createHostTypeTriggerQuestion = exports.createHostTypeTriggerQuestionForVS = exports.HostTypeTriggerOptionsForVS = exports.HostTypeTriggerOptions = exports.AppServiceOptionItemForVS = exports.AppServiceOptionItem = exports.FunctionsHttpTriggerOptionItem = exports.FunctionsTimerTriggerOptionItem = void 0;
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const featureFlags_1 = require("../../../common/featureFlags");
const localizeUtils_1 = require("../../../common/localizeUtils");
const question_1 = require("../../solution/fx-solution/question");
const constants_1 = require("./constants");
const strings_1 = require("./resources/strings");
// NOTE: id must be the sample as cliName to prevent parsing error for CLI default value.
exports.FunctionsTimerTriggerOptionItem = optionWithL10n({
    id: "timer-functions",
    hostType: strings_1.HostTypes.AZURE_FUNCTIONS,
    trigger: strings_1.NotificationTriggers.TIMER,
});
exports.FunctionsHttpTriggerOptionItem = optionWithL10n({
    id: "http-functions",
    hostType: strings_1.HostTypes.AZURE_FUNCTIONS,
    trigger: strings_1.NotificationTriggers.HTTP,
});
exports.AppServiceOptionItem = optionWithL10n({
    id: "http-restify",
    hostType: strings_1.HostTypes.APP_SERVICE,
    // trigger of app service host is hard-coded to http, so no need to set here
});
// TODO: this option will not be shown in UI, leave messages empty.
exports.AppServiceOptionItemForVS = {
    id: "http-webapi",
    hostType: strings_1.HostTypes.APP_SERVICE,
    label: "",
    cliName: "",
    description: "",
    detail: "",
};
exports.HostTypeTriggerOptions = [
    exports.AppServiceOptionItem,
    exports.FunctionsHttpTriggerOptionItem,
    exports.FunctionsTimerTriggerOptionItem,
];
exports.HostTypeTriggerOptionsForVS = [exports.AppServiceOptionItemForVS];
function createHostTypeTriggerQuestionForVS() {
    const prefix = "plugins.bot.questionHostTypeTrigger";
    return {
        name: constants_1.QuestionNames.BOT_HOST_TYPE_TRIGGER,
        title: localizeUtils_1.getLocalizedString(prefix + ".title"),
        type: "singleSelect",
        staticOptions: exports.HostTypeTriggerOptionsForVS,
        default: exports.AppServiceOptionItemForVS.id,
        placeholder: localizeUtils_1.getLocalizedString(prefix + ".placeholder"),
        skipSingleOption: true,
        validation: {
            validFunc: async (input) => {
                const name = input;
                if (name.length === 0) {
                    return localizeUtils_1.getLocalizedString(`${prefix}.error.emptySelection`);
                }
                return undefined;
            },
        },
    };
}
exports.createHostTypeTriggerQuestionForVS = createHostTypeTriggerQuestionForVS;
// The restrictions of this question:
//   - appService and function are mutually exclusive
//   - users must select at least one trigger.
function createHostTypeTriggerQuestion(platform) {
    const prefix = "plugins.bot.questionHostTypeTrigger";
    let staticOptions;
    if (platform === teamsfx_api_1.Platform.CLI) {
        // The UI in CLI is different. It does not have description. So we need to merge that into label.
        staticOptions = exports.HostTypeTriggerOptions.map((option) => {
            // do not change the original option
            const cliOption = Object.assign({}, option);
            cliOption.label = `${option.label} (${option.description})`;
            return cliOption;
        });
    }
    else {
        staticOptions = exports.HostTypeTriggerOptions;
    }
    return {
        name: constants_1.QuestionNames.BOT_HOST_TYPE_TRIGGER,
        title: localizeUtils_1.getLocalizedString(`${prefix}.title`),
        type: "multiSelect",
        staticOptions: staticOptions,
        default: [exports.AppServiceOptionItem.id],
        placeholder: localizeUtils_1.getLocalizedString(`${prefix}.placeholder`),
        validation: {
            validFunc: async (input) => {
                const name = input;
                if (name.length === 0) {
                    return localizeUtils_1.getLocalizedString(`${prefix}.error.emptySelection`);
                }
                if (name.includes(exports.AppServiceOptionItem.id) && name.length > 1) {
                    return localizeUtils_1.getLocalizedString(`${prefix}.error.hostTypeConflict`);
                }
                return undefined;
            },
        },
        onDidChangeSelection: async function (currentSelectedIds, previousSelectedIds) {
            if (currentSelectedIds.size > 1 && currentSelectedIds.has(exports.AppServiceOptionItem.id)) {
                if (previousSelectedIds.has(exports.AppServiceOptionItem.id)) {
                    currentSelectedIds.delete(exports.AppServiceOptionItem.id);
                }
                else {
                    currentSelectedIds = new Set([exports.AppServiceOptionItem.id]);
                }
            }
            return currentSelectedIds;
        },
    };
}
exports.createHostTypeTriggerQuestion = createHostTypeTriggerQuestion;
// Question model condition to determine whether to show "Select triggers" question after "Select capabilities".
// Return undefined for true, a string for false. The string itself it not used.
exports.showNotificationTriggerCondition = {
    validFunc: (input, inputs) => {
        if (!inputs) {
            return "Invalid inputs";
        }
        if (featureFlags_1.isPreviewFeaturesEnabled()) {
            const cap = inputs[question_1.AzureSolutionQuestionNames.Capabilities];
            if (cap === question_1.NotificationOptionItem.id) {
                return undefined;
            }
            // Single Select Option for "Add Feature"
            const feature = inputs[question_1.AzureSolutionQuestionNames.Features];
            if (feature === question_1.NotificationOptionItem.id) {
                return undefined;
            }
        }
        else {
            const cap = inputs[question_1.AzureSolutionQuestionNames.Capabilities];
            if (Array.isArray(cap) && cap.includes(question_1.NotificationOptionItem.id)) {
                return undefined;
            }
        }
        return "Notification is not selected";
    },
    // Workaround for CLI: it requires containsAny to be set, or it will crash.
    containsAny: [question_1.NotificationOptionItem.id],
};
function optionWithL10n(option) {
    // e.g. expands to plugins.bot.triggers.functionsTimer.label
    const prefix = "plugins.bot.triggers";
    return Object.assign(Object.assign({}, option), { label: localizeUtils_1.getLocalizedString(`${prefix}.${option.id}.label`), cliName: localizeUtils_1.getLocalizedString(`${prefix}.${option.id}.cliName`), description: localizeUtils_1.getLocalizedString(`${prefix}.${option.id}.description`), detail: localizeUtils_1.getLocalizedString(`${prefix}.${option.id}.detail`) });
}
