// @microsoft/teamsfx-core/build/plugins/resource/appstudio/utils/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertToAppDefinitionMessagingExtensions = exports.convertToAppDefinitionBots = exports.convertToAppDefinition = exports.getLocalAppName = exports.getCustomizedKeys = exports.checkAndConfig = exports.replaceConfigValue = void 0;
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
const constants_1 = require("../constants");
const localizeUtils_1 = require("../../../../common/localizeUtils");
const results_1 = require("../results");
function replaceConfigValue(config, id, value) {
    if (config && id && value) {
        const idTag = `{${id}}`;
        while (config.includes(idTag)) {
            config = config.replace(idTag, value);
        }
    }
    return config;
}
exports.replaceConfigValue = replaceConfigValue;
/**
 *
 * @throws Error - when placeholder doesn't have corresponding value
 */
function checkAndConfig(config, id, value) {
    const idTag = `{${id}}`;
    if (value) {
        return replaceConfigValue(config, id, value);
    }
    else {
        if (config.includes(idTag)) {
            throw results_1.AppStudioResultFactory.SystemError("RequiredDataMissing", [
                localizeUtils_1.getDefaultString("plugins.appstudio.dataRequired", idTag),
                localizeUtils_1.getLocalizedString("plugins.appstudio.dataRequired", idTag),
            ]);
        }
        else {
            return config;
        }
    }
}
exports.checkAndConfig = checkAndConfig;
function getCustomizedKeys(prefix, manifest) {
    let keys = [];
    for (const key in manifest) {
        if (manifest.hasOwnProperty(key)) {
            const value = manifest[key];
            if (typeof value === "object") {
                if (Array.isArray(value)) {
                    value.map((item, index) => {
                        keys = keys.concat(getCustomizedKeys(`${prefix}.${key}[${index}]`, item));
                    });
                }
                else {
                    keys = keys.concat(getCustomizedKeys(`${prefix}.${key}`, value));
                }
            }
            else if (typeof value === "string" && value.startsWith("{{config.manifest")) {
                keys.push(`${prefix}.${key}`);
            }
        }
    }
    return keys;
}
exports.getCustomizedKeys = getCustomizedKeys;
function getLocalAppName(appName) {
    const suffix = "-local-debug";
    if (suffix.length + appName.length <= constants_1.TEAMS_APP_SHORT_NAME_MAX_LENGTH) {
        appName = appName + suffix;
    }
    return appName;
}
exports.getLocalAppName = getLocalAppName;
/**
 * Convert from TeamsAppManifest to AppDefinition
 * Localization file is not included
 * @param appManifest
 * @returns
 */
function convertToAppDefinition(appManifest) {
    const appDefinition = {
        appName: appManifest.name.short,
        validDomains: appManifest.validDomains,
    };
    appDefinition.showLoadingIndicator = appManifest.showLoadingIndicator;
    appDefinition.isFullScreen = appManifest.isFullScreen;
    appDefinition.appId = appManifest.id;
    appDefinition.appName = appManifest.name.short;
    appDefinition.shortName = appManifest.name.short;
    appDefinition.longName = appManifest.name.full;
    appDefinition.manifestVersion = appManifest.manifestVersion;
    appDefinition.version = appManifest.version;
    appDefinition.packageName = appManifest.packageName;
    appDefinition.accentColor = appManifest.accentColor;
    appDefinition.developerName = appManifest.developer.name;
    appDefinition.mpnId = appManifest.developer.mpnId;
    appDefinition.websiteUrl = appManifest.developer.websiteUrl;
    appDefinition.privacyUrl = appManifest.developer.privacyUrl;
    appDefinition.termsOfUseUrl = appManifest.developer.termsOfUseUrl;
    appDefinition.shortDescription = appManifest.description.short;
    appDefinition.longDescription = appManifest.description.full;
    appDefinition.staticTabs = appManifest.staticTabs;
    appDefinition.configurableTabs = appManifest.configurableTabs;
    appDefinition.bots = convertToAppDefinitionBots(appManifest);
    appDefinition.messagingExtensions = convertToAppDefinitionMessagingExtensions(appManifest);
    appDefinition.connectors = appManifest.connectors;
    appDefinition.devicePermissions = appManifest.devicePermissions;
    if (appManifest.webApplicationInfo) {
        appDefinition.webApplicationInfoId = appManifest.webApplicationInfo.id;
        appDefinition.webApplicationInfoResource = appManifest.webApplicationInfo.resource;
    }
    appDefinition.activities = appManifest.activities;
    if (appManifest.icons.color) {
        appDefinition.colorIcon = appManifest.icons.color;
    }
    if (appManifest.icons.outline) {
        appDefinition.outlineIcon = appManifest.icons.outline;
    }
    return appDefinition;
}
exports.convertToAppDefinition = convertToAppDefinition;
function convertToAppDefinitionBots(appManifest) {
    const bots = [];
    if (appManifest.bots) {
        appManifest.bots.forEach((manBot) => {
            var _a, _b, _c;
            const teamCommands = [];
            const groupCommands = [];
            const personalCommands = [];
            (_a = manBot === null || manBot === void 0 ? void 0 : manBot.commandLists) === null || _a === void 0 ? void 0 : _a.forEach((list) => {
                list.commands.forEach((command) => {
                    teamCommands.push({
                        title: command.title,
                        description: command.description,
                    });
                    groupCommands.push({
                        title: command.title,
                        description: command.description,
                    });
                    personalCommands.push({
                        title: command.title,
                        description: command.description,
                    });
                });
            });
            const bot = {
                botId: manBot.botId,
                isNotificationOnly: (_b = manBot.isNotificationOnly) !== null && _b !== void 0 ? _b : false,
                supportsFiles: (_c = manBot.supportsFiles) !== null && _c !== void 0 ? _c : false,
                scopes: manBot.scopes,
                teamCommands: teamCommands,
                groupChatCommands: groupCommands,
                personalCommands: personalCommands,
            };
            bots.push(bot);
        });
    }
    return bots;
}
exports.convertToAppDefinitionBots = convertToAppDefinitionBots;
function convertToAppDefinitionMessagingExtensions(appManifest) {
    const messagingExtensions = [];
    if (appManifest.composeExtensions) {
        appManifest.composeExtensions.forEach((ext) => {
            var _a;
            const me = {
                botId: ext.botId,
                canUpdateConfiguration: true,
                commands: ext.commands,
                messageHandlers: (_a = ext.messageHandlers) !== null && _a !== void 0 ? _a : [],
            };
            messagingExtensions.push(me);
        });
    }
    return messagingExtensions;
}
exports.convertToAppDefinitionMessagingExtensions = convertToAppDefinitionMessagingExtensions;
